### ISSUE TITLE
Unable to return a grpc::ClientContext from a function.

### ISSUE THREAD

**Issue by tonicsoft** [grpc/grpc#16680]
### What version of gRPC and what language are you using?
 1.15, C++
 
### What operating system (Linux, Windows, …) and version?
 Ubuntu 18.04
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 gcc 7.3.0
 
### What did you do?
try to return a ClientContext from a function, to avoid repetitive code on every call:
```
grpc::ClientContext context_factory() {
    grpc::ClientContext context;
    return std::move(context);
}

ClientContext my_ctx(std::move(context_factory()));
```
 
### What did you expect to see?
 ClientContext move constructor invoked
 
### What did you see instead?
fails to compile due to private copy constructor. Either there should be a move constructor defined, or it should be explicitly deleted with a comment to say why this object is not movable.

**Comment by yashykt** [comment]
why do you want to do something like this? We explicitly disallow copy and assign for clientcontext

**Comment by tonicsoft** [comment]
A new client context must be created for every call. If any customisation
is required (for example, setting a timeout) the creation code becomes
duplicated in many places and should be factored out to a factory method.

I understand why copy is disallowed, however, I see no reason it shouldn't
be moved to facilitate it to be returned from said factory method.

**Comment by coryan** [comment]
Ping.  This would also be useful for us at googleapis/google-cloud-cpp.

**Comment by bduclaux** [comment]
Ping also. The only workaround is to use a macro to avoid rewriting the same code in each RPC client function.

**Comment by coryan** [comment]
Please do not automatically cleanup this without at least telling us if this is a valid request.

**Comment by coryan** [comment]
Hello?  Anybody there?

**Comment by coryan** [comment]
Are you hoping that the bugs will go away if you ignore them long enough?

**Comment by yashykt** [comment]
Sorry for the trouble! We've recently disabled the stale bot for kind/bug issues

**Comment by sxlijin** [comment]
My workaround for this (which is not the _best_ since it does change allocation semantics) is to simply return a `std::unique_ptr<grpc::ClientContext>` from a factory function instead of using a macro.

**Comment by yashykt** [comment]
Checking in - We ended up adding a `Mutex` inside `ClientContext` which is going to make this trickier. This might go away in the future, but unsure.

Currently, `std::unique_ptr<grpc::ClientContext>` is the workaround.

**Comment by markdroth** [comment]
I don't think this is something we want to do.  Applications can either return `std::unique_ptr<grpc::ClientContext>` as mentioned above, or they can have a function that takes the `ClientContext` by reference and populates its fields rather than returning it.
